### PR TITLE
MOD-4252: Coordinator leak fix

### PR DESCRIPTION
**Pull request by GuyAv46**
This PR fixes a leak found on the coordinator when we have an error on a query that contains VSS.
This PR fixes [MOD-4252](https://redislabs.atlassian.net/browse/MOD-4252)

### CODE REVIEW

**Review by DvirDukhan — DISMISSED** [review]

**Comment by oshadmi** [comment]
Leak reported in [test_vecsim:test_timeout_reached](https://app.circleci.com/pipelines/github/RediSearch/RediSearch/10626/workflows/b792d733-dd28-404b-b46d-56cf235774e6/jobs/52222/steps?invite=true#step-109-6042)

```
=================================================================
==150417==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 1920024 byte(s) in 1 object(s) allocated from:
    #0 0x51c7fd in __interceptor_malloc (/usr/local/bin/redis-server-asan-6.2+0x51c7fd)
    #1 0x587aba in ztrymalloc_usable /tmp/redis.0m7hyc8f/redis/src/zmalloc.c:101:17
    #2 0x587aba in zmalloc /tmp/redis.0m7hyc8f/redis/src/zmalloc.c:119:17
    #3 0x6f8be8 in RM_Alloc /tmp/redis.0m7hyc8f/redis/src/module.c:436:12
    #4 0x7f0eb4f41d68 in rm_malloc /root/project/src/rmalloc.h:12:10
    #5 0x7f0eb4f46ded in searchResultReducer /root/project/coord/src/module.c:1213:28
    #6 0x7f0eb4f92700 in fanoutCallback /root/project/coord/src/rmr/rmr.c:195:7
    #7 0x7f0eb5493729 in __redisRunCallback /root/project/deps/hiredis/async.c:293:9
    #8 0x7f0eb5491c43 in redisProcessCallbacks /root/project/deps/hiredis/async.c:576:13
    #9 0x7f0eb54bbe65 in redisSSLAsyncRead /root/project/deps/hiredis/ssl.c:528:9
    #10 0x7f0eb5493bb4 in redisAsyncHandleRead /root/project/deps/hiredis/async.c:666:5
    #11 0x7f0eb4f7dce2 in redisLibuvPoll /root/project/deps/hiredis/adapters/libuv.h:22:9
    #12 0x7f0eb548899e in uv__io_poll /root/project/deps/libuv/src/unix/epoll.c:374:11

SUMMARY: AddressSanitizer: 1920024 byte(s) leaked in 1 allocation(s).
```

**Review by oshadmi — APPROVED** [review]

File: src/util/heap.c [hunk 1]
```diff
@@ -58,6 +58,14 @@ void heap_free(heap_t *h) {
   rm_free(h);
 }
 
+// Useful when you want to free all the internal data
+void heap_destroy(heap_t *h) {
+  for (size_t i = 0; i < h->count; i++) {
+    rm_free(h->array[i]);
+  }
+  rm_free(h);
```
Comment (oshadmi) [inline src/util/heap.c:66]: Consider `heap_free(h)` (although we still use `rm_free` for the array data)

**Review by oshadmi — DISMISSED** [review]

File: coord/src/module.c [hunk 2]
```diff
@@ -1247,11 +1246,12 @@ static int searchResultReducer(struct MRCtx *mc, int count, MRReply **replies) {
...
-    while ((sr = heap_poll(rCtx.pq))) {
-      rm_free(sr);
-    }
-    heap_free(rCtx.pq);
+    heap_destroy(rCtx.pq);
+  }
+  if (rCtx.reduceSpecialCaseCtx &&
+      rCtx.reduceSpecialCaseCtx->specialCaseType == SPECIAL_CASE_KNN &&
+      rCtx.reduceSpecialCaseCtx->knn.pq) {
+    heap_destroy(rCtx.reduceSpecialCaseCtx->knn.pq);
```
Comment (ashtul) [inline coord/src/module.c:1254]:
```suggestion
  if (rCtx.reduceSpecialCaseCtx->knn.pq) {
    heap_destroy(rCtx.reduceSpecialCaseCtx->knn.pq);
```

**Review by ashtul — COMMENTED** [review]
Can we add a test that will mimick the leak w/o the fix?

File: coord/src/module.c (hunk 2 above)
Comment (GuyAv46) [reply coord/src/module.c:1254]: Why can I skip checking these conditions?

Comment (GuyAv46) [reply coord/src/module.c:1254]: WRT adding a test: I'm not sure how to reproduce this. It was a flake timeout test that depended on the stage at which the timeout occurred, the leak could or could not happen.

Comment (ashtul) [reply coord/src/module.c:1254]: @GuyAv46 is there a case where `rCtx.reduceSpecialCaseCtx->knn.pq` exists and should not be freed?

Comment (GuyAv46) [reply coord/src/module.c:1254]: No, but what if `rCtx.reduceSpecialCaseCtx` was not initialized? Or is initialized not on special case KNN (the heap is a part of a union)

**Review by ashtul — APPROVED** [review]
LGTM

**Pull request merged**